auslib/db.py

```diff
@@ -2098,6 +2096,30 @@ def _proceedIfNotReadOnly(self, name, limit=None, transaction=None):
         if self.isReadOnly(name, limit, transaction):
             raise ReadOnlyError("Release '%s' is read only" % name)
 
+    def change_readonly(self, where, is_readonly, changed_by, old_data_version, transaction=None, dryrun=False, signoffs=None):
+        release = self.select(where=where, columns=[self.name, self.product], transaction=transaction)[0]
+        product = release["product"]
+
+        if is_readonly:
+            if not self.db.hasPermission(changed_by, "release_read_only", "set", product, transaction):
+                raise PermissionDeniedError(f"{changed_by} is not allowed to mark {product} products read only")
+        else:
+            if not self.db.hasPermission(changed_by, "release_read_only", "unset", product, transaction):
+                raise PermissionDeniedError(f"{changed_by} is not allowed to mark {product} products read write")
+
+        if not dryrun and not is_readonly:
+
+            def _getPotentialRequiredSignoffs(required_signoffs):
+                return [obj for v in required_signoffs for obj in v]
+
+            potential_required_signoffs = _getPotentialRequiredSignoffs(self.getPotentialRequiredSignoffs([release], transaction=transaction).values())
+            if not potential_required_signoffs:
+                potential_required_signoffs = _getPotentialRequiredSignoffs(
+                    self.getPotentialRequiredSignoffsForProduct(release["product"], transaction=transaction).values()
```

Comment:
I want to make sure I fully understand this section. It looks to me like the process here is:
* We check for regular Required Signoffs (ie: is the Release pointed at by a Rule that requires signoff)
* If it is, we use those signoffs
* Otherwise, use the sum of all of the signoffs for the Release's product

Is that right? Either way, this section could use an explanatory comment, as it's not obvious why would we potentially check for two different types of Required Signoffs.

Reply:
It is a sum of all roles and MAX(signoffs) for them, i.e.:

Product|Channel|Role|Signoffs Required
----------|----------|------|-----------------------
Firefox|Release|relman|1
Firefox|Release|releng|1
Firefox|Beta|releng|2
Firefox|Beta|other_role|1

Signoffs required for product:
relman: 1
other_role: 1
releng: 2

Reply:
Right, yes - sorry. The max for each role, not the sum. But my first bullet (about using regular Required Signoffs) is correct, right? That's what it looks like from reading the code and local testing.

Reply:
Yeah, that is right. First we look for associated rule, if it is null, we get signoffs for product.